x/programs/examples/imports/program/program.go

```diff
@@ -170,31 +162,34 @@ func (i *Import) callProgramFn(
 	return int64(res[0])
 }
 
-func getCallArgs(ctx context.Context, memory runtime.Memory, buffer []byte, invokeProgramID int64) ([]int64, error) {
+func getCallArgs(ctx context.Context, memory runtime.Memory, buffer []byte, programIDBytes []byte) ([]int64, error) {
 	// first arg contains id of program to call
-	args := []int64{invokeProgramID}
-	p := codec.NewReader(buffer, len(buffer))
-	i := 0
-	for !p.Empty() {
-		size := p.UnpackInt64(true)
-		isInt := p.UnpackBool()
-		if isInt {
-			valueInt := p.UnpackInt64(true)
-			args = append(args, valueInt)
-		} else {
-			valueBytes := make([]byte, size)
-			p.UnpackFixedBytes(int(size), &valueBytes)
-			ptr, err := runtime.WriteBytes(memory, valueBytes)
-			if err != nil {
-				return nil, err
-			}
-			args = append(args, ptr)
-		}
-		i++
+	bytes := imports.PrependLength(programIDBytes)
+	invokeProgramIDPtr, err := runtime.WriteBytes(memory, bytes)
+	if err != nil {
+		return nil, err
 	}
-	if p.Err() != nil {
-		return nil, fmt.Errorf("failed to unpack arguments: %w", p.Err())
+
+	args := []int64{invokeProgramIDPtr}
+
+	for i := 0; i < len(buffer); {
+		// unpacks uint32
+		lenBytes := buffer[i : i+consts.Uint32Len]
+		length := binary.BigEndian.Uint32(lenBytes) + consts.Uint32Len
+
+		// we include the length in the value bytes because the program
+		// still needs to know how many bytes to read
+		valueBytes := buffer[i : i+int(length)]
+		i += int(length)
+
+		// every argument is a pointer
+		ptr, err := runtime.WriteBytes(memory, valueBytes)
```

Comment:
Would it be more efficient to write once and then set offsets into the pointer?

Reply:
(or is that not really a good idea in this context)

Reply:
yea that could work, i think what you are thinking of is similar to this issue I just created #638